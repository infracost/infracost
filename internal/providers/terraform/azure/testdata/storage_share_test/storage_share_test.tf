provider "azurerm" {
  skip_provider_registration = true
  features {}
}

resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "westus"
}

locals {
  share_premium_options = [
    {
      access_tiers              = ["Premium"]
      account_replication_types = ["LRS", "ZRS"],
    },
  ]

  share_standard_options = [
    {
      access_tiers              = ["Cool", "Hot", "TransactionOptimized"]
      account_replication_types = ["LRS", "GRS", "ZRS"],
    },
  ]

  storage_account_premium_permutations = distinct(flatten([
    for share_premium_option in local.share_premium_options : [
      for account_replication_type in share_premium_option.account_replication_types : {
        account_replication_type = account_replication_type
      }
    ]
  ]))

  share_premium_permuations = distinct(flatten([
    for share_premium_option in local.share_premium_options : [
      for account_replication_type in share_premium_option.account_replication_types : [
        for access_tier in share_premium_option.access_tiers : {
          access_tier              = access_tier
          account_replication_type = account_replication_type
        }
      ]
    ]
  ]))

  storage_account_standard_permutations = distinct(flatten([
    for share_standard_option in local.share_standard_options : [
      for account_replication_type in share_standard_option.account_replication_types : {
        account_replication_type = account_replication_type
      }
    ]
  ]))

  share_standard_permuations = distinct(flatten([
    for share_standard_option in local.share_standard_options : [
      for account_replication_type in share_standard_option.account_replication_types : [
        for access_tier in share_standard_option.access_tiers : {
          access_tier              = access_tier
          account_replication_type = account_replication_type
        }
      ]
    ]
  ]))
}

resource "azurerm_storage_account" "filestorage" {
  for_each = { for entry in local.storage_account_premium_permutations : "${entry.account_replication_type}" => entry }

  name                     = substr(lower("filestorage${each.value.account_replication_type}"), 0, 24)
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location
  account_kind             = "FileStorage"
  account_tier             = "Premium"
  account_replication_type = each.value.account_replication_type
}

resource "azurerm_storage_share" "premium-filestorage" {
  for_each = { for entry in local.share_premium_permuations : "${entry.access_tier}.${entry.account_replication_type}" => entry }

  name                 = substr(lower("premium${each.value.access_tier}${each.value.account_replication_type}"), 0, 24)
  storage_account_name = azurerm_storage_account.filestorage["${each.value.account_replication_type}"].name
  quota                = 50
  access_tier          = each.value.access_tier
}

resource "azurerm_storage_account" "standard" {
  for_each = { for entry in local.storage_account_standard_permutations : "${entry.account_replication_type}" => entry }

  name                     = substr(lower("standard${each.value.account_replication_type}"), 0, 24)
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location
  account_kind             = "StorageV2"
  account_tier             = "Standard"
  account_replication_type = each.value.account_replication_type
}

resource "azurerm_storage_share" "standard" {
  for_each = { for entry in local.share_standard_permuations : "${entry.access_tier}.${entry.account_replication_type}" => entry }

  name                 = substr(lower("standard${each.value.access_tier}${each.value.account_replication_type}"), 0, 24)
  storage_account_name = azurerm_storage_account.standard["${each.value.account_replication_type}"].name
  quota                = 50
  access_tier          = each.value.access_tier
}

resource "azurerm_storage_account" "premium" {
  name                     = "premium"
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location
  account_kind             = "StorageV2"
  account_tier             = "Premium"
  account_replication_type = "LRS"
}

resource "azurerm_storage_share" "unsupported" {
  name                 = "unsupported"
  storage_account_name = azurerm_storage_account.premium.name
  quota                = 50
  access_tier          = "Premium"
}
