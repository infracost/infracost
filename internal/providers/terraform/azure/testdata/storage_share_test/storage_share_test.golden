
 Name                                                               Monthly Qty  Unit                      Monthly Cost 
                                                                                                                        
 azurerm_storage_account.filestorage["LRS"]                                                                             
 ├─ Data at rest                                             Monthly cost depends on usage: $0.18 per GB                
 └─ Snapshots                                                Monthly cost depends on usage: $0.15 per GB                
                                                                                                                        
 azurerm_storage_account.filestorage["ZRS"]                                                                             
 ├─ Data at rest                                             Monthly cost depends on usage: $0.22 per GB                
 └─ Snapshots                                                Monthly cost depends on usage: $0.19 per GB                
                                                                                                                        
 azurerm_storage_account.premium                                                                                        
 ├─ Capacity                                                 Monthly cost depends on usage: $0.20 per GB                
 ├─ Write operations                                         Monthly cost depends on usage: $0.0228 per 10k operations  
 ├─ List and create container operations                     Monthly cost depends on usage: $0.065 per 10k operations   
 ├─ Read operations                                          Monthly cost depends on usage: $0.00182 per 10k operations 
 └─ All other operations                                     Monthly cost depends on usage: $0.00182 per 10k operations 
                                                                                                                        
 azurerm_storage_account.standard["GRS"]                                                                                
 ├─ Capacity                                                 Monthly cost depends on usage: $0.0458 per GB              
 ├─ Write operations                                         Monthly cost depends on usage: $0.11 per 10k operations    
 ├─ List and create container operations                     Monthly cost depends on usage: $0.11 per 10k operations    
 ├─ Read operations                                          Monthly cost depends on usage: $0.0044 per 10k operations  
 ├─ All other operations                                     Monthly cost depends on usage: $0.0044 per 10k operations  
 └─ Blob index                                               Monthly cost depends on usage: $0.069 per 10k tags         
                                                                                                                        
 azurerm_storage_account.standard["LRS"]                                                                                
 ├─ Capacity                                                 Monthly cost depends on usage: $0.0208 per GB              
 ├─ Write operations                                         Monthly cost depends on usage: $0.055 per 10k operations   
 ├─ List and create container operations                     Monthly cost depends on usage: $0.055 per 10k operations   
 ├─ Read operations                                          Monthly cost depends on usage: $0.0044 per 10k operations  
 ├─ All other operations                                     Monthly cost depends on usage: $0.0044 per 10k operations  
 └─ Blob index                                               Monthly cost depends on usage: $0.039 per 10k tags         
                                                                                                                        
 azurerm_storage_account.standard["ZRS"]                                                                                
 ├─ Capacity                                                 Monthly cost depends on usage: $0.026 per GB               
 ├─ List and create container operations                     Monthly cost depends on usage: $0.06875 per 10k operations 
 ├─ Read operations                                          Monthly cost depends on usage: $0.0044 per 10k operations  
 ├─ All other operations                                     Monthly cost depends on usage: $0.0044 per 10k operations  
 └─ Blob index                                               Monthly cost depends on usage: $0.039 per 10k tags         
                                                                                                                        
 azurerm_storage_share.premium-filestorage["Premium.LRS"]                                                               
 ├─ Data at rest                                                             50  GB                               $8.80 
 └─ Snapshots                                                            10,000  GB                           $1,500.00 
                                                                                                                        
 azurerm_storage_share.premium-filestorage["Premium.ZRS"]                                                               
 ├─ Data at rest                                                             50  GB                              $11.00 
 └─ Snapshots                                                            10,000  GB                           $1,870.00 
                                                                                                                        
 azurerm_storage_share.standard["Cool.GRS"]                                                                             
 ├─ Data at rest                                                      2,000,000  GB                         $100,200.00 
 ├─ Snapshots                                                            20,000  GB                           $1,002.00 
 ├─ Metadata at rest                                                     20,000  GB                           $1,310.00 
 ├─ Read operations                                                          20  10k operations                   $0.26 
 ├─ Write operations                                                        200  10k operations                  $52.00 
 ├─ List operations                                                         200  10k operations                  $28.60 
 ├─ Other operations                                                        200  10k operations                   $1.14 
 └─ Data retrieval                                                        2,000  GB                              $20.00 
                                                                                                                        
 azurerm_storage_share.standard["Cool.LRS"]                                                                             
 ├─ Data at rest                                                      2,000,000  GB                          $45,600.00 
 ├─ Snapshots                                                            20,000  GB                             $456.00 
 ├─ Metadata at rest                                                     20,000  GB                             $594.00 
 ├─ Read operations                                                          20  10k operations                   $0.26 
 ├─ Write operations                                                        200  10k operations                  $26.00 
 ├─ List operations                                                         200  10k operations                  $14.30 
 ├─ Other operations                                                        200  10k operations                   $1.14 
 └─ Data retrieval                                                        2,000  GB                              $20.00 
                                                                                                                        
 azurerm_storage_share.standard["Cool.ZRS"]                                                                             
 ├─ Data at rest                                                      2,000,000  GB                          $57,000.00 
 ├─ Snapshots                                                            20,000  GB                             $570.00 
 ├─ Metadata at rest                                                     20,000  GB                             $742.00 
 ├─ Read operations                                                          20  10k operations                   $0.26 
 ├─ Write operations                                                        200  10k operations                  $26.00 
 ├─ List operations                                                         200  10k operations                  $17.88 
 ├─ Other operations                                                        200  10k operations                   $1.14 
 └─ Data retrieval                                                        2,000  GB                              $20.00 
                                                                                                                        
 azurerm_storage_share.standard["Hot.GRS"]                                                                              
 ├─ Data at rest                                                      2,000,000  GB                         $126,400.00 
 ├─ Snapshots                                                            20,000  GB                           $1,264.00 
 ├─ Metadata at rest                                                     20,000  GB                           $1,310.00 
 ├─ Read operations                                                          20  10k operations                   $0.11 
 ├─ Write operations                                                        200  10k operations                  $28.60 
 ├─ List operations                                                         200  10k operations                  $28.60 
 └─ Other operations                                                        200  10k operations                   $1.14 
                                                                                                                        
 azurerm_storage_share.standard["Hot.LRS"]                                                                              
 ├─ Data at rest                                                      2,000,000  GB                          $57,400.00 
 ├─ Snapshots                                                            20,000  GB                             $574.00 
 ├─ Metadata at rest                                                     20,000  GB                             $594.00 
 ├─ Read operations                                                          20  10k operations                   $0.11 
 ├─ Write operations                                                        200  10k operations                  $14.30 
 ├─ List operations                                                         200  10k operations                  $14.30 
 └─ Other operations                                                        200  10k operations                   $1.14 
                                                                                                                        
 azurerm_storage_share.standard["Hot.ZRS"]                                                                              
 ├─ Data at rest                                                      2,000,000  GB                          $72,000.00 
 ├─ Snapshots                                                            20,000  GB                             $720.00 
 ├─ Metadata at rest                                                     20,000  GB                             $742.00 
 ├─ Read operations                                                          20  10k operations                   $0.11 
 ├─ Write operations                                                        200  10k operations                  $17.88 
 ├─ List operations                                                         200  10k operations                  $17.88 
 └─ Other operations                                                        200  10k operations                   $1.14 
                                                                                                                        
 azurerm_storage_share.standard["TransactionOptimized.GRS"]                                                             
 ├─ Data at rest                                                      2,000,000  GB                         $200,000.00 
 ├─ Snapshots                                                            20,000  GB                           $2,000.00 
 ├─ Read operations                                                          20  10k operations                   $0.03 
 ├─ Write operations                                                        200  10k operations                   $6.00 
 ├─ List operations                                                         200  10k operations                   $3.00 
 └─ Other operations                                                        200  10k operations                   $0.30 
                                                                                                                        
 azurerm_storage_share.standard["TransactionOptimized.LRS"]                                                             
 ├─ Data at rest                                                      2,000,000  GB                         $120,000.00 
 ├─ Snapshots                                                            20,000  GB                           $1,200.00 
 ├─ Read operations                                                          20  10k operations                   $0.03 
 ├─ Write operations                                                        200  10k operations                   $3.00 
 ├─ List operations                                                         200  10k operations                   $3.00 
 └─ Other operations                                                        200  10k operations                   $0.30 
                                                                                                                        
 azurerm_storage_share.standard["TransactionOptimized.ZRS"]                                                             
 ├─ Data at rest                                                      2,000,000  GB                         $150,000.00 
 ├─ Snapshots                                                            20,000  GB                           $1,500.00 
 ├─ Read operations                                                          20  10k operations                   $0.03 
 ├─ Write operations                                                        200  10k operations                   $3.75 
 ├─ List operations                                                         200  10k operations                   $3.00 
 └─ Other operations                                                        200  10k operations                   $0.30 
                                                                                                                        
 OVERALL TOTAL                                                                                              $946,944.87 
──────────────────────────────────
19 cloud resources were detected:
∙ 17 were estimated, 6 of which include usage-based costs, see https://infracost.io/usage-file
∙ 1 was free:
  ∙ 1 x azurerm_resource_group
∙ 1 is not supported yet, see https://infracost.io/requested-resources:
  ∙ 1 x azurerm_storage_share

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┓
┃ Project                                            ┃ Monthly cost ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━┫
┃ TestStorageShare                                   ┃ $946,945     ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━┛
Logs:

level=warning msg="Skipping resource azurerm_storage_share.unsupported. Premium access tier is only supported for FileStorage accounts"